### DISCUSSION TITLE
Difference between API and flutter/dart Package

### DISCUSSION THREAD

**Discussion by MartinForgetProfessionnal** [openfoodfacts/openfoodfacts-dart#721] [General]
Hi! i was first using the API and then realised that there was a package for flutter/dart, so i made the switch, but i noticed some differences and one of them i can't figure how to resolve the problem it causes me.
I use nutrients values per serving, but in some cases the mesurement unit differt from mg to g for exemple... In my first version, working with the API i had no problem because i had the option tto display salt_unit. With the package, the getter is deprecated and it's replaced with typicalUnit which doesn't vary to fit the value per serving that is returned. The thing is that i really don't want to make 2 request to fetch that unic missing information.
So my question is... Can i get that unit (dynamicly) somehow using the package?

**Comment by monsieurtanuki** [comment]
Hi @MartinForgetProfessionnal!

If needed I can go back to the different comments in closed issues, but I think that the rule is that "salt" is in grams: `salt(typicalUnit: Unit.G, offTag: 'salt')`.
So to answer your question, you can get the unit, it's `Nutrient.salt.typicalUnit`. Always.
Whatever value you get (and set) should be in grams.
Of course there may be cases where the display is strange: in that case you can convert in your app in milligrams - I'm sure you can do the math for the conversion ;)

> With the package, the getter is deprecated

Btw the getter does not exist anymore, since last December I think.

Please keep us posted if you find discrepancies regarding values and units, between the API and dart.

**Comment by MartinForgetProfessionnal** [comment]
![SaltExample](https://user-images.githubusercontent.com/107161274/228911534-14daf180-8d35-4b4e-b37c-4a66ae39f639.JPG)
![saltExample2](https://user-images.githubusercontent.com/107161274/228911537-9e1e7c3b-1fa5-40f4-9c5b-ded540b50e07.JPG)

**Reply by monsieurtanuki** [reply]
To me, that means that:
* for 100g, there is 1.4275 grams of salt
* for a serving, there is .4 grams of salt
* last time someone set the salt value, it was in `mg` (but we store the value in grams anyway)

If you have a look at 5411188115472 you'll find vitamin B12, that is normally (in the real world) displayed in micrograms:
* `"vitamin-b12":3.8e-07` // obviously in grams
* `"vitamin-b12_100g":3.8e-07` // obviously in grams
* `"vitamin-b12_unit":"µg"` // typed in in micrograms, e.g. on the website

**Reply by MartinForgetProfessionnal** [reply]
Ok, happy to ear it from you. 
In resume, if i stick with
UnitHelper.unitToString(Nutrient.(salt, fat, calcium, etc...).typicalUnit), 
i'm sure to be ok.

**Reply by MartinForgetProfessionnal** [reply]
One more question, i really want to do things right. if the unit is returned as mg with the UnitHelper, the value you store is mg rightt? Or i do need to make the conversion (ex: calcium, iron...)

**Reply by monsieurtanuki** [reply]
If it's a weight, the value is _always_ in grams.
The `typicalUnit` is a good enough approximation of the nutrient unit (e.g. µg for B12), marginally different from what you read in some product packages.
If you want to display typical unit and a value, you have to convert the value from grams to the typical unit. Same for saving: you save in grams, and should convert from whatever unit the user sees.

**Reply by MartinForgetProfessionnal** [reply]
Happy i asked that last question, i tought the opposite. Thanks alot! i really really appreciate your time! have a good one.

**Comment by MartinForgetProfessionnal** [comment]
that is the part that gets me messing me up...

**Reply by MartinForgetProfessionnal** [reply]
the same is true for some example with sodium. It also diverge in some cases with calcium and iron (% or mg)
